### ISSUE TITLE
Problem with evaluation abstractions

### ISSUE THREAD

**Issue by jwiegley** [haskell-nix/hnix#75]
This is needed by @ryantrinkle.

https://github.com/jwiegley/hnix/blob/master/Nix/Eval.hs#L193

**Comment by madjar** [comment]
Progress report: I've been working on this for a few weeks.

The core idea is to change NVFunction to
```
NVFunction (Params (ValueSet m -> m r)) (ValueSet m -> m r)
```

This allows to pass the argumentset to itself at application time, to evaluate default values.

The current status is that is works when evaluated in `Identity`, but freezes (maybe with an infinite loop?) in `IO`. I still have to trace stuff to figure out what's happening here.

If anybody's curious, the WIP code is here: https://github.com/jwiegley/hnix/compare/master...madjar:fix-recursive-argument-set

**Comment by jwiegley** [comment]
@madjar Which test are you using to reproduce the freeze?

**Comment by madjar** [comment]
@jwiegley I added a small reproducing example (in the middle in Eval.hs, to make debugging easier :))

```
expr = let Success x = parseNixString "({ x ? 1, y ? x * 3 }: y - x) {}"
       in x
goeval :: MonadFix m => m (NValue m)
goeval = evalExpr expr mempty

goevalIO = goeval @ IO
goevalId = goeval @ Identity
```

`goevalId` returns `(NVConstant (NInt 2)))`, `goevalIO` freezes.

**Comment by jwiegley** [comment]
@madjar I've looked at this with @ryantrinkle, and it looks like we're running into a problem with laziness.  It works with Identity, because nothing is getting forced, and it just so happens then that Nix's laziness gets implemented by Haskell's own laziness.

In IO, however, there are strict evaluation boundaries, which violate Nix's laziness, resulting in cycles that can't be broken, hence the bad MVar dependency.

I think the fix is to reify the concept of laziness and make it a value, using a new NVThunk data type, with the same semantics as Haskell/Nix. Thus, buildArgument would just stick the pending evaluation into an NVThunk, but wouldn't force it until it's actually used.  For Identity this will likely require using IORef and `unsafePerformIO` to allow fixing the thunk contents to avoid re-evaluation.

Does this sound like anything you'd be interested in tackling?

**Comment by madjar** [comment]
@jwiegley I've reached a similar conclusion (though not as articulate, as you can see here: https://github.com/jwiegley/hnix/issues/67#issuecomment-369343772).

I guess I started working on something along the lines of what you describe, replacing `ValueSet` with `type Environment m = Text -> m (Maybe (NValue m))` (which is equivalent to `Map Text (m (NValue m))`, except you can't list keys ; see note [1]). But I'm not sure this is what you are suggesting.

Anyway, I'm definitely interested, but I feel I am not understanding everything here. Could you write a little bit more about your idea of NVThunk to get me started?


Note [1]: On top of the recursiveness of `buildArgument`, there is one more problem: in `evalBinds`, the keys of the set can be dynamic and depend on the set itself, which makes it impossible to list the keys without evaluation the whole set (thus making `Map` not usable).

**Comment by Ericson2314** [comment]
Could we interleave the Monad into the value type, and then do `pure . unsafePerformIO` in the IO case to continue leveraging built-in laziness in all cases?

(On can also manually write something like `pure . unsafePerformIO` but merging the returned inner state token instead of dropping it which might help maintain the order of effects.)

**Comment by ryantrinkle** [comment]
I think that might be very hard to extend to other monads (e.g.
transformers over io)

**Comment by madjar** [comment]
Progress report: I tried going with the `NVThunk` idea, but I kind of reached a wall. I'm not sure I quite groked it, and I'd love some hints on where `NVThunk` would actually fit.

**Comment by jwiegley** [comment]
I'll need to carve some time to check this out. Basically we want to do a sort of Löb transform: instead of seeing attribute sets as mappings of the type `Map Key Value`, they are really mappings from keys to functions, of type `Map Key (Map Key Value -> Value)`.  If `NVThunk` were a kind of Value, then we'd be building a thunk that references the parent map, allowing us to return back to `Map Key Value`, where now some of the values are effectively deferred functions of type `Map Key Value -> Value`.  Part of the complexity in our case comes from the presence of a monad during evaluation, such that `m` needs to be added in all the necessary places above.

**Comment by jwiegley** [comment]
OK! This is finally fixed, and in the right way (not by avoiding IO, since that made supporting `import` in a principled way unpleasant).

The crux of the problem we were having can be demonstrated by this test case:

``` haskell
module Bug where

import           Control.Monad.Fix
import qualified Data.Map.Lazy as Map
import           Data.Maybe
import           Debug.Trace

loebM :: (MonadFix m, Traversable t) => t (t a -> m a) -> m (t a)
loebM f = mfix $ \a -> trace "mapM" $ mapM (trace "$ a" . ($ a)) f

main :: IO ()
main = do
    xs <- loebM $ Map.fromList
        [ (0, const (pure (2 :: Int)))
        , (1, \l ->
                  trace ("before: l = " ++ show (Map.keys l))
                    $ pure (fromJust
                            (trace
                                ("after: l = " ++ show (Map.keys l))
                                  $ Map.lookup 0 l) + 3)) ]
    traceM "loebM completed"
    print xs
```

If you remove the "before" trace, it works; if you leave it in, it dies. The reason here is what we had suspected before: monadic binds introduce sequence points. In order for `loebM` to do its job in the way that we want, the call to `mapM` has to be completed before we make any references to the members of the map. Otherwise, the `l` argument doesn't exist yet.

The answer for `hnix` is to add one more layer of indirection. `ValueSet` has changed from:

    type ValueSet m = Map.Map Text (NValue m)

to:

    type ValueSet m = Map.Map Text (m (NValue m))

This injection of an `m` layer allows us to defer monadic evaluation of the members of the map until after all the wiring has been done by `loeb` (we can now use regular `loeb` instead of `loebM`, since we don't want the effects processed during construction of the self-referential map).

Another change I made to show why our monadic layering is useful is a new `MonadNix` type class:

``` haskell
class MonadFix m => MonadNix m where
    currentScope :: m (ValueSet m)
    newScope :: ValueSet m -> m r -> m r
    importFile :: NValue m -> m (NValue m)
```

This basically represents all of the effects that occur during evaluation of a Nix expression: defining and referencing dynamic attribute scopes, and importing files. I'm sure more will be added in future.

This type class also simplified a lot of the code, because now instead of building lambdas everywhere, we can just use `currentScope` and `newScope`, which have much clearer meaning in the contexts where they are used.

Here's the IO definition of `MonadNix`. Note the use of a helper type `Cyclic` to represent the fact that our monadic effects might occur anywhere throughout the evaluation tree:

``` haskell
newtype Cyclic m a = Cyclic { runCyclic :: StateT (ValueSet (Cyclic m)) m a }
    deriving (Functor, Applicative, Monad, MonadFix)

instance MonadNix (Cyclic IO) where
    currentScope = Cyclic get
    newScope s k = Cyclic $ put s >> runCyclic k
    importFile path = Cyclic $ case path of
        Fix (NVLiteralPath path) -> do
            eres <- parseNixFile path
            case eres of
                Failure err  -> error $ "Parse failed: " ++ show err
                Success expr -> runCyclic $ evalExpr expr
        _ -> error $ "Unexpected argument to import: " ++ show path
```

And last but not least, I added a helper function that came up during ICFP 2017, and which Conal Elliott (@conal) helped me to discover: `adiM`. This implements abstract definitional interpreters for catamorphisms over functor fixed points. The basic result is that, in the same way that recursion schemes let us arbitrarily layer in new data (such as parsing annotations), abstract definitional interpreters let us layer in new behavior on top of existing evaluators. This made it trivial to create a new tracing evaluator, which the tests use so that you can see where evaluation went wrong:

``` haskell
tracingExprEval :: MonadNix m => NExpr -> IO (m (NValue m))
tracingExprEval =
    fmap (runIdentity . snd) . adiM @() (pure <$> eval) psi
  where
    psi k v@(Fix x) = do
        putStrLn $ "Evaluating: " ++ show x
        k v
```

Ping @ryantrinkle: You should be unblocked now, and have a way to play with imports as you wish!

**Comment by jwiegley** [comment]
Note that this exact same issue is currently plaguing all the other uses of `rec` in `Eval.hs` too, which can probably be resolved in a similar way. This is causing, for example, this test to fail:

```
git submodule update --init
cp ./data/nix/tests/lang/lib.nix .
./dist/build/hnix/hnix --debug --eval -f /Users/johnw/src/hnix/data/nix/tests/lang/eval-okay-list.nix
```

**Comment by jwiegley** [comment]
OK, that last problem should be sorted now as well, by the same approach.

**Comment by jwiegley** [comment]
I was able to remove the need for `MonadFix` entirely, using just `loeb` when recursive references are desired, and I've deferred management of the "scope stack" to the `MonadNix` instances, rather than dealing with it explicitly in the evaluator. At the moment it's implemented by merging maps, but we should move to a more efficient method since the depth will never be all that large, but the number of attributes will be.

**Comment by jwiegley** [comment]
Thanks to @tgiannak, we were able to discern that the whole of this issue could be resolved by simply adding a `MonadInterleave` constraint, which is `id` for non-strict monads, like `Identity`, and uses `unsafeInterleaveIO` for `IO`. The only requires is that we have a way to indicate that certain binds (those where actions need to be deferred) can be evaluated lazily.

**Comment by Ericson2314** [comment]
@jwiegley yeah that's similar but better than what I proposed above.

I believe we can improve it still further though. The problem with lazy IO is forcing a pure value can have side effect. We don't actually need that but rather the opposite as the problem is caching as without `unsafeInterleave` we have `call-by-name` instead of `call-by-need`.

If we instead do
```haskell
class MonadInterleave m where
    maybeSafeInterleave :: m a -> m (m a)
```
we have it so the binding the inner monadic value forces the effectful thunk and (as before) binding the outer value signals the earliest the effect can happen. No pure value comes with side effects.

For the `IO` implementation, we do
```haskell
maybeSafeInterleave = fmap Control.Exception.Base evaluate . unsafeInterleaveIO
```
`evaluate` isn't perfect as we just want to evaluate the `State#` and not actual WHNF, but I doubt it will be a problem in practice.

**Comment by jwiegley** [comment]
@Ericson2314 Do you have an example of a case where the difference introduced by `maybeSafeInterleave` is needed?  Switching to `MonadInterleave` simplified the code but had no effect on any of the tests, so it seemed like it offered the semantics we need. But maybe there are tests still breaking that need what you're suggesting. Can you clarify?

**Comment by jwiegley** [comment]
@Ericson2314 Note that the type of the evaluator we're using is effectively:

``` haskell
NExprF (m (NValue m)) -> m (NValue m)
```

So when evaluating an `NIf`, for example, we bind the condition, but then only bind the chosen alternative.

Where interleaving comes into play is when we're building up the values of a set or a function argument set, where we only want strictness in the spine and not the leaves. Then we use `interleave` on the monadic actions provided by the earlier phases of the evaluator, so that we can reference the values they *will* compute to, when needed later on.

**Comment by Ericson2314** [comment]
@jwiegley I'll take a look at the code. But in between replying and reading, I stumbled on this Oleg paper http://homes.soic.indiana.edu/ccshan/rational/lazy-nondet.pdf which, while not using `unsafeInterleaveIO` or anything like it, does have a `share :: m a -> m (m a)` i.e analogous signature.

**Comment by jwiegley** [comment]
@Ericson2314 That's a great paper, I'll put that on my short list. Thank you for thinking more deeply about this issue, you've convinced me that this needs further consideration before I make more refactoring changes. Reopening this issue until we've resolved the question.
